### PR TITLE
fix explain in function_score if no function filter matches

### PR DESCRIPTION
**Pull request by brwe**
When each function in function_score has a filter but none of them matches
we always assume 1 for the combined functions and then combine that with the
sub query score.
But the explanation did not reflect that because in case no function matched
we did not even use the actual score that was computed in the explanation.

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/common/lucene/search/function/FiltersFunctionScoreQuery.java [hunk 1]
```diff
@@ -224,17 +224,23 @@ public Explanation explain(LeafReaderContext context, int doc) throws IOExceptio
...
+                factorExplanation = Explanation.match(
                         CombineFunction.toFloat(score),
                         "function score, score mode [" + scoreMode.toString().toLowerCase(Locale.ROOT) + "]",
                         filterExplanations);
-                expl = combineFunction.explain(expl, factorExplanation, maxBoost);
+
+            } else {
+                // it is a little weird to add a match although no function matches but that is the way function_score behaves right now
+                factorExplanation = Explanation.match(1.0f,
+                    "No function matched", new ArrayList<>());
```
Comment (danielmitterdorfer) [inline core/src/main/java/org/elasticsearch/common/lucene/search/function/FiltersFunctionScoreQuery.java]: Nit: you could use `Collections.emptyList()` here.

File: core/src/test/java/org/elasticsearch/index/query/functionscore/FunctionScoreTests.java [hunk 2]
```diff
@@ -726,6 +726,31 @@ public void testFilterFunctionScoreHashCodeAndEquals() {
...
+        CombineFunction combineFunction = randomFrom(new
+            CombineFunction[]{CombineFunction.SUM, CombineFunction.AVG, CombineFunction.MIN, CombineFunction.MAX,
+            CombineFunction.MULTIPLY, CombineFunction.REPLACE});
+        
+        // check for document that has no macthing function
+        FiltersFunctionScoreQuery query = new FiltersFunctionScoreQuery(new TermQuery(new Term(FIELD, "out")), scoreMode,
+            new FilterFunction[]{new FilterFunction(new TermQuery(new Term("_uid", "2")), new WeightFactorFunction(10))},
+            Float.MAX_VALUE, Float.NEGATIVE_INFINITY, combineFunction);
+        TopDocs searchResult = localSearcher.search(query, 1);
+        Explanation exp1 = localSearcher.explain(query, searchResult.scoreDocs[0].doc);
```
Comment (danielmitterdorfer) [inline core/src/test/java/org/elasticsearch/index/query/functionscore/FunctionScoreTests.java]: Nit: remove the 1 in `exp1` as there is only one instance in the test?

**Comment by danielmitterdorfer** [comment]
Left two very minor comments. Otherwise LGTM.

**Comment by brwe** [comment]
@danielmitterdorfer thanks a lot for the review! addressed all comments.

**Pull request merged**